File: Lernwerk/Services/Review/SpacedRepetition.swift
import Foundation

enum ReviewGrade: Int, CaseIterable {
    case again = 0
    case hard = 3
    case good = 4
    case easy = 5

    var label: String {
        switch self {
        case .again: return "Nochmal"
        case .hard: return "Schwer"
        case .good: return "Gut"
        case .easy: return "Leicht"
        }
    }
}

struct SchedulingState: Equatable {
    static let initialEase = 2.5
    static let minimumEase = 1.3
    static let new = SchedulingState(intervalDays: 0, easeFactor: initialEase, repetitions: 0, lapses: 0)

    var intervalDays: Int
    var easeFactor: Double
    var repetitions: Int
    var lapses: Int
}

/// SM-2 scheduling; a failed card comes back after a short relearning delay instead of the next day.
enum SpacedRepetition {
    static let relearnDelay: TimeInterval = 10 * 60

    static func next(_ state: SchedulingState, grade: ReviewGrade) -> SchedulingState {
        var result = state
        let quality = Double(grade.rawValue)
        let easeChange = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
        result.easeFactor = max(SchedulingState.minimumEase, state.easeFactor + easeChange)

        if grade == .again {
            result.repetitions = 0
            result.intervalDays = 0
            result.lapses += 1
            return result
        }

        result.repetitions += 1
        switch result.repetitions {
        case 1:
            result.intervalDays = 1
        case 2:
            result.intervalDays = 6
        default:
            result.intervalDays = max(1, Int((Double(state.intervalDays) * result.easeFactor).rounded()))
        }
        return result
    }

    static func dueDate(for state: SchedulingState, reviewedAt date: Date, calendar: Calendar = .current) -> Date {
        guard state.intervalDays > 0 else {
            return date.addingTimeInterval(relearnDelay)
        }
        let startOfDay = calendar.startOfDay(for: date)
        return calendar.date(byAdding: .day, value: state.intervalDays, to: startOfDay) ?? date
    }
}

File: android/app/src/main/java/de/maxifrz/lernwerk/review/SpacedRepetition.kt
package de.maxifrz.lernwerk.review

import kotlinx.serialization.Serializable
import java.time.Instant
import java.time.ZoneId
import kotlin.math.roundToInt

enum class ReviewGrade(val quality: Int, val label: String) {
    AGAIN(0, "Nochmal"),
    HARD(3, "Schwer"),
    GOOD(4, "Gut"),
    EASY(5, "Leicht"),
}

@Serializable
data class SchedulingState(
    val intervalDays: Int = 0,
    val easeFactor: Double = INITIAL_EASE,
    val repetitions: Int = 0,
    val lapses: Int = 0,
) {
    companion object {
        const val INITIAL_EASE = 2.5
        const val MINIMUM_EASE = 1.3
        val NEW = SchedulingState()
    }
}

/** SM-2 scheduling; a failed card comes back after a short relearning delay instead of the next day. */
object SpacedRepetition {
    const val RELEARN_DELAY_SECONDS = 10L * 60

    fun next(state: SchedulingState, grade: ReviewGrade): SchedulingState {
        val quality = grade.quality.toDouble()
        val easeChange = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
        val ease = maxOf(SchedulingState.MINIMUM_EASE, state.easeFactor + easeChange)

        if (grade == ReviewGrade.AGAIN) {
            return state.copy(easeFactor = ease, repetitions = 0, intervalDays = 0, lapses = state.lapses + 1)
        }

        val repetitions = state.repetitions + 1
        val interval = when (repetitions) {
            1 -> 1
            2 -> 6
            else -> maxOf(1, (state.intervalDays * ease).roundToInt())
        }
        return state.copy(easeFactor = ease, repetitions = repetitions, intervalDays = interval)
    }

    fun dueDate(state: SchedulingState, reviewedAt: Instant, zone: ZoneId = ZoneId.systemDefault()): Instant {
        if (state.intervalDays <= 0) return reviewedAt.plusSeconds(RELEARN_DELAY_SECONDS)
        return reviewedAt.atZone(zone).toLocalDate().plusDays(state.intervalDays.toLong()).atStartOfDay(zone).toInstant()
    }
}
